net: getaddrinfo: Query both IPv4 and IPv6 if hints is missing

jukkar opened the pull request:
We must query both IPv4 and IPv6 addresses if the hints parameter
is NULL i.e., user does not supply hints.

Fixes #16453

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: getaddrinfo: Query both IPv4 and IPv6 if family is AF_UNSPEC

We must query both IPv4 and IPv6 addresses if the hints parameter
is NULL i.e., user does not supply hints or if family is set to
AF_UNSPEC.

Fixes #16453

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
> net: getaddrinfo: Query both IPv4 and IPv6 if hints is missing

Nitpick about commit message: fairly speaking, the matter should be that IPv4 and IPv6 should be both queried if hints has AF_UNSPEC address family (and that's default if hints is NULL at all).

pfalcon commented on subsys/net/lib/sockets/getaddrinfo.c:
```diff
@@ -226,6 +226,23 @@ int z_impl_z_zsock_getaddrinfo_internal(const char *host, const char *service,
 		st1 = DNS_EAI_SYSTEM;
 	}
 
+	if (family == AF_UNSPEC && IS_ENABLED(CONFIG_NET_IPV6)) {
```
Would be nice to get a comment here what exactly happens here (i.e. that if family == AF_UNSPEC, we already have queried IPv4 case in the previous block).

pfalcon commented on subsys/net/lib/sockets/getaddrinfo.c:
```diff
@@ -105,7 +105,7 @@ static int exec_query(const char *host, int family,
 	enum dns_query_type qtype = DNS_QUERY_TYPE_A;
 
-	if (IS_ENABLED(CONFIG_NET_IPV6) && family != AF_INET) {
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
```
Probably worth adding a comment, that for AF_UNSPEC case, we query IPv4 in this func.

pfalcon reviewed (commented):
Thanks for picking this up. To avoid future confusion when working with this func, would be nice to add one, better both, of: a) code comments; b) testcase for this behavior.

As you know, I never insist on b), though that's a perfect example where we need this... (Again, not insisting.)

jukkar committed:
tests: net: getaddrinfo: Make sure we receive two queries

If user supplies AF_UNSPEC, we need to do two queries, one for
IPv4 A record and one for IPv6 AAAA record.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Updated according to comments.

pfalcon reviewed (approved):
Thanks for the updates, looks great!

Removing platform_whitelist was a bit too brave and hits the same issue patch from @andrewboie hit earlier. Better to skip that change for now I guess.

jukkar commented:
> Removing platform_whitelist was a bit too brave and hits the same issue patch from @andrewboie hit earlier. Better to skip that change for now I guess.

I have been removing the whitelists in various PRs. Created an issue yesterday for this one #16778

jukkar commented:
Rebased against latest master.

pfalcon commented:
As this no longer contains any changes not described in the topic of the PR, I'd suggest to merge it. (I guess @tbursztyka is on vacation, nor this is his known area of interest.)

jukkar commented:
recheck
